### DISCUSSION TITLE
Will the model binding be deprecated?

### DISCUSSION THREAD

**Discussion by lairg99** [livewire/livewire#6909] [4. General]
Some of my applications rely heavily on v2's legacy model binding and I have appreicated this feature although there are a few things to keep in mind. 

However, as I'm preparing to upgrade to v3, I'm wondering if the model binding will be deprecated in the future, as the description in the config file talks about a "feature flag", while "legacy" could describe the initiation of deprecation.

What is desired in the future? Can we continue to rely on the feature? If so, perhaps the parameter should be renamed from `legacy_model_binding` to `model_binding` or `features.model_binding`.

**Comment by ju5t** [comment]
I would not rely on this in the future. To be fair, I'm not against deprecating this in the future either in v4 for example. You can always bring it back yourself if it's really needed. We will actively refactor to array-based data though (combined with Spatie's data package).

**Reply by lairg99** [reply]
I don't get the advantage of array-based data, while with a model everything is nicely typed. For me, it always feels like duplicate code. Do I miss something? 🧐

For example, currently I would implemented something like this:

```php
class Update extends Component
{
    public Post $post;

    protected function rules(): array {
        return [
            'post.title' => ['required', 'string'],
            'post.content' => ['required', 'string'],
        ];
    }
    
    public function save() {
        $this->validate();
        
        $this->post->save();
    }
}
```

But, without model binding - in the simplest way - I have to implement something like this:

```php
class Update extends Component
{
    public Post $post;

    public string $title = '';
    public string $content = '';

    public function mount() {
        $this->title = $this->post->title;
        $this->content = $this->post->content;
    }

    protected function rules(): array {
        return [
            'title' => ['required', 'string'],
            'content' => ['required', 'string'],
        ];
    }

    public function save() {
        $this->validate();

        $this->post->fill([
            'title' => $this->title,
            'content' => $this->content
        ])->save();
    }
}
```

Where is the benefit? I really appreciate the easy and short solution. Although we have a larger application, it's maintainable. Sure, you can extract some of this logic into form classes - but in my eyes, it's over-engineered.

**Reply by ju5t** [reply]
There are a few caveats with model binding. Most important, security. If you are not careful, all properties of a model are visible & editable through the frontend. That's not great.

It has been discussed here before v3 was released. If I remember correctly, most users were OK with slashing support for models. Partly because it's easy to bring back through synths if absolutely necessary.

Personally, I'm ok with it. We will be shifting to Spatie's Data package, it's a shame v3 didn't go that route but it is what it is.

**Reply by lairg99** [reply]
Thank you for your thoughts. I'll dive into it. May you could give some insights of your integration with Spatie's DTOs?

**Reply by thumbtech** [reply]
Related to this, I remain a little confused about the ongoing support for model properties in V3. If I can simply mount a model instance into a property (independent of `legacy_model_binding` and yes with same security concerns as V2) then what has actually been deprecated?

The call-out regarding "[Eloquent constraints aren't preserved between requests](https://livewire.laravel.com/docs/properties#eloquent-constraints-arent-preserved-between-requests)" and the discussion of computed properties in V3 are really a brain fart for me, implying that the JSON flow does not include the data, but rather somehow only the query?!? Maybe V2 was the same and I didn't realize it.

**Reply by joshhanley** [reply]
@thumbtech in V2 and V3, the way Livewire handles Eloquent models/ Eloquent collections assigned to properties, is to store the ID/s and the class name of the model/collection. That is all that is sent to the frontend. On rehydration, Livewire makes a database call to "restore" the models by getting them fresh out of the database.

The difference between V2 and V3, is V3 doesn't allow you to bind to Eloquent models/ Eloquent collections, as this requires model attributes to be sent to the front end (which has security implications) and a significant amount of code complexity. There were issues with the V2 implementation when people tried to bind to custom queries (like using custom selects, joins, etc.).

Due to this and other reasons, we decided that in V3, you can still assign an Eloquent model/ Eloquent collection to a property for convenience, so you don't need to deal with it, but you can no longer bind to them (using `wire:model`).

For compatibility reasons and to make upgrading easier, we implemented `legacy_model_binding`, which is meant to replicate how V2 handled model binding (although it may have some bugs), to allow people who were using model binding to upgrade, and refactor later.

We don't recommend using `legacy_model_binding` going forward, hence why it's disabled by default, to encourage all new development to use properties on a component or the new form objects.

So regarding the ongoing support, in V3 assigning Eloquent models/ Eloquent collections to properties will be supported. Model binding (using `wire:model` to a model property) is currently supported through `legacy_model_binding`, we have no plans to remove this anytime soon, but we do not recommend using it, unless you need it to upgrade an existing app. If for some reason we do decide to remove the `legacy_model_binding` in the future, developers will still be able to copy the Synths to their own app, and they can maintain them going forward.

Hope this helps!

**Reply by joshhanley** [reply]
@lairg99 hopefully the above answers your questions too. If you have any others, feel free to ask 🙂

**Reply by lairg99** [reply]
Thank you for your response! I will take a look.

**Reply by ju5t** [reply]
Something to consider: v3 `legacy_model_binding` doesn't use your model `$casts`. It uses the built-in synths. You may have to write some of your own synths for unknown data types (we had to).

**Reply by thumbtech** [reply]
FWIW I would suggest that the data binding limitations of a model property are not obvious in the current [properties docs](https://livewire.laravel.com/docs/properties). A simple asterisk in [the list of common PHP types](https://livewire.laravel.com/docs/properties#common-php-types) with a footnote of "* Two-way data binding **not** supported" would be helpful.

Another minor V3 docs quibble is that examples manipulating data are consistently of the Post::create() variety and best practice(s) for an efficient & secure update is not obvious until, e.g., one wanders down into [don't trust property values](https://livewire.laravel.com/docs/properties#dont-trust-property-values). Notice the little documentation shortcut taken between [a V2 save](https://laravel-livewire.com/docs/2.x/properties#binding-models) and [a V3 save](https://livewire.laravel.com/docs/actions) (save converted to create in the so-called save method)? V3 user is left thinking OK but how could I actually Post::save() in V3 ... where to store my Post, etc.?

But for sure: massive kudos to overall incredible package & docs to the whole livewire team!

**Reply by shokanshi** [reply]
> The difference between V2 and V3, is V3 doesn't allow you to bind to Eloquent models/ Eloquent collections, as this requires model attributes to be sent to the front end (which has security implications) and a significant amount of code complexity. There were issues with the V2 implementation when people tried to bind to custom queries (like using custom selects, joins, etc.).

Hi @joshhanley, does this mean in V2, model attributes are sent to the front end and they have security implications? Knowing the answer is important as it will affect the decision on migration existing projects to V3.

**Reply by joshhanley** [reply]
@shokanshi not by default, no. If you want to bind to any model attributes, you have to add the attributes you want to bind to your $rules array property for Livewire to send that value to the front end. That's the same in V2 and legacy model binding in V3.

**Reply by johannesobier** [reply]
In Caleb's screen casts it's shown how to make a Form class. One of the member variables was a public model object (User). And for every used field there was an own member variable. Then there was a method setUser($user) that sets' the User model in the Form object and also the field member variables. When saving, the values of the field member variables are copied to the corresponding model member and then the model object was saved. That's such an enormous amount of boilerplate code! With or without a Form class.  I've several applications with heavily use of model binding. And don't know how to sell my customers why to pay me for work that hasn't any obvious improvement. And the upgrade wouldn't be an easy task, because in some of those projects I've used a lot of inheritance, Livewire component base classes with derivations. So I'm not happy with this decision.

**Reply by lairg99** [reply]
A little hope for you: As it sounds, we used Livewire in a similar way. Recently we upgraded to v3 and with model binding enabled everything seemed to work.

What we noticed, however, is that the model binding with wire:model.live is lagging. When typing, the text is sometimes replaced by an older version (possibly by the response). If you type slowly, everything works.

**Reply by ju5t** [reply]
And should they remove it in v4 or later, you can reimplement it yourself using synths. The code is here:
https://github.com/livewire/livewire/tree/main/src/Features/SupportLegacyModels

**Comment by thumbtech** [comment]
> ... So I'm not happy with this decision.

I had to upgrade a large-ish app relying on LW2 model inheritance, too. My models had some custom geospatial types that had busted when used in LW2, so I already had a taste of the complicated object juggling that was occurring and (mostly) working well under the hood in LW2. I think they've made the right call to move away from all that.

I have some wide tables and just could not see separate props for every field, so I handled upgrade like this:

LW2 component handling create or update of Post model:

```
use App\Post;
 
class PostForm extends Component
{
    public Post $post;
 
    protected $rules = [
        'post.title' => 'required|string|min:6',
        'post.content' => 'required|string|max:500',
    ];
 
    public function save()
    {
        $this->validate();
 
        $this->post->save();
    }

    public function mount(Post $post)
    {
        $this->post = $post;
    }
}
```

Converted to LW3:

```
use App\Post;
 
class PostForm extends Component
{
    public Post $postModel;

    public $post; // array of data you will bind to in the blade, populated in mount function!

    protected $fields = [
        'title',
        'content',
    ]; // the fields from model you will be editing!
 
    protected $rules = [
        'post.title' => 'required|string|min:6',
        'post.content' => 'required|string|max:500',
    ];
 
    public function save()
    {
        $this->post = array_slice_assoc($this->post, $this->fields); // filter out any injected $post vars!

        $this->validate();
 
        // create or update?
        if ($this->postModel->id)
        {
            $this->postModel->update($this->post);
        }
        else
        {
            $this->postModel = Post::create($this->post);
        }

        session()->flash('message', "\"{$this->postModel->title}\" successfully " . ($this->postModel->wasRecentlyCreated ? 'created!' : 'updated!'));
    }

    public function mount(Post $postModel)
    {
        $this->postModel = $postModel;
        $this->post = $this->postModel->only($this->fields);
    }
}
```

I changed the prior $post model property to $postModel and then let $post be a simple array. That way I didn't have to mess with the blade or validation too much.

FWIW that array_slice_assoc function is a helper I made from some code on php.net. There is probably a better way to do that!

```
if (!function_exists('array_slice_assoc'))
{
    function array_slice_assoc(array $array, array $keys)
    {
        return array_intersect_key($array, array_flip($keys));
    }
}
```

**Comment by johannesobier** [comment]
One thing I stumbled upon after uprade to LW3 with legacy_model_binding is, I had to cast the model's boolean/tinyint(1) attributes to boolean, to have proper support for <input type="checkbox" ...>:
protected $casts [
    'tit' => 'boolean',
    'tat' => 'boolean',
    ...
];
Took me hours...

**Reply by ju5t** [reply]
On this topic, I found that some custom casts on a model required custom Livewire synths too. A bit unexpected, as I thought it use the model cast instead.

**Reply by sbkrogers** [reply]
WOW! Nice find.

**Comment by ChrisSantiago82** [comment]
I really hope the support of model binding will continue in future versions.
And what would be the recommended way to this in version 3 if there is for example a large model to edit?
To create variables from all model fields and when you want to save it the other way around, doesn't seem to be a very clean way. A model could be turned into an array.

Either way, the code won't be that clean anymore.

**Reply by lairg99** [reply]
You can do everything as in v2. If it is no longer supported, you can use your own [synths](https://livewire.laravel.com/docs/synthesizers). See [Model Synth](https://github.com/livewire/livewire/blob/main/src/Features/SupportLegacyModels/EloquentModelSynth.php).

**Reply by thumbtech** [reply]
FWIW the [array method](https://github.com/livewire/livewire/discussions/6909#discussioncomment-7908711) is working well for me. Too much scary magic in that EloquentModelSynth class! KISS is king!

**Reply by joshhanley** [reply]
@ChrisSantiago82 at this stage we have no plans to remove the legacy model support but we strongly suggest people don't use it for future development as we may remove it in a future major version.

But as @lairg99 suggested, if that happens you can just copy the legacy EloquentModelSynth and legacy EloquentCollectionSynth into your own project and maintain them yourself indefinitely. 🙂 https://github.com/livewire/livewire/tree/main/src/Features/SupportLegacyModels

**Comment by ChrisSantiago82** [comment]
Ok, thanks for the information

**Reply by joshhanley** [reply]
@ChrisSantiago82 no worries! 🙂
